sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -258,37 +258,73 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       val dsOptions = new CaseInsensitiveStringMap(options.asJava)
 
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
-      provider.getTable(dsOptions) match {
-        case table: SupportsWrite if table.supports(BATCH_WRITE) =>
-          if (partitioningColumns.nonEmpty) {
-            throw new AnalysisException("Cannot write data to TableProvider implementation " +
-              "if partition columns are specified.")
+      mode match {
+        case SaveMode.Append | SaveMode.Overwrite =>
+          val table = provider match {
+            case supportsExtract: SupportsCatalogOptions =>
```

Comment:
Could this be simplified to just call `saveAsTable` (or an equivalent internal method with the right args) when `SupportsCatalogOptions` is implemented? Or are the behaviors different?

Reply:
drop table gets a bit weird in overwrite mode.

Reply:
You're right, `saveAsTable` uses RTAS instead of truncate.